designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,469 @@
+
+## Documentation
+
+- It needs an entry in the migration guide.
+- The "[Node.js API](https://eslint.org/docs/developer-guide/nodejs-api)" page should describe the new public API and deprecation of `CLIEngine` class.
+
+## Drawbacks
+
+People that use `CLIEngine` have to update their application with the new API. It will need hard work.
+
+## Backwards Compatibility Analysis
+
+This RFC is a drastic change, but not a breaking change until we decide to remove `CLIEngine` class.
+
+This RFC just adds `LinterShell` class and deprecates `CLIEngine` class. We can do both in a minor release.
+
+## Alternatives
+
+### Alternatives for the new class
+
+Adding `CLIEngine#executeOnFilesAsync()` method is an alternative.
+
+**Pros:**
+
+- It's smaller change than adding `LinterShell` class.
+
+**Cons:**
+
+- We have to maintain both synchronous and asynchronous APIs. It's kind of hard works.
+  - We can deprecate the synchronous version, but in that case, I feel odd because `executeOnFiles()` is the essential name of `executeOnFilesAsync()`.
+- We need the asynchronous version of the other APIs in the future. The `getFormatterAsync()` is needed for formatters with progress. The asynchronous version of the other methods is needed for ES modules.
+  - `CLIEngine` will get huge because we cannot share the most code between asynchronous API and synchronous API.
+  - API users need the number of API migrations.
+- It causes confusion for API users. As Node.js built-in libraries are so, I guess that people expect the two (`executeOnFiles()` and `executeOnFilesAsync()`) to have exactly the same features. However, it's impossible. We have a bundle of the features that we cannot support on synchronous API, such as linting in parallel, formatters with progress, ES modules, etc.
+
+Therefore, I think that introducing the new class that has only asynchronous API makes more sense. This RFC solves:
+
+- We can freeze the code of the synchronous version of our API by deprecating `CLIEngine`. This reduces the maintenance cost of duplicate codes.
+- We can reduce the number of API migrations for API users.
+- We can reduce the confusion of similar but different methods.
+- And as a bonus, we can reduce the confusion of the name of `CLIEngine`.
+
+### Alternatives for disallow execution in parallel
+
+- Throwing if the previous call has not finished yet (fail-fast).
+- Aborting the previous call then run (steal ownership of the cache file).
+
+are alternatives. Both cases stop the previous or current call. It may be surprising users.
+
+- Waiting the previous call internally.
+
+The access of the cache file finishes regardless of the progress of the iterator that the method returned. It means that API users cannot know when the file access finished. On the other hand, because `LinterShell` objects know when the file access finished, it can execute the next call at the proper timing.
```

Comment:
I'm confused by the first sentence here. Which method returns an iterator? Or should this say `regardless of the progress of the promise that the method returned`?

Reply:
Oh, I forgot to update this section. I will